src/main/java/com/khutircraftubackend/userprofile/UserProfileService.java

```diff
@@ -0,0 +1,72 @@
+package com.khutircraftubackend.userprofile;
+
+import com.khutircraftubackend.auth.AuthenticationService;
+import com.khutircraftubackend.auth.UserEntity;
+import com.khutircraftubackend.auth.UserRepository;
+import com.khutircraftubackend.auth.UserService;
+import com.khutircraftubackend.mail.EmailSender;
+import com.khutircraftubackend.userprofile.request.ChangeOfEmailRequest;
+import com.khutircraftubackend.userprofile.request.ConfirmEmailRequest;
+import com.khutircraftubackend.userprofile.response.ChangeOfEmailResponse;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+
+import java.security.Principal;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.Map;
+
+@RequiredArgsConstructor
+@Setter
+@Getter
+public class UserProfileService {
+
+    private final UserService userService;
+    private final UserRepository userRepository;
+    private final AuthenticationService authenticationService;
+    private final EmailSender emailSender;
+    private Map<String, EmailTokenInfo> confirmEmail;
+
+    public ChangeOfEmailResponse changeOfEmail(Principal principal, ChangeOfEmailRequest request) {
+        UserEntity user = userService.findByPrincipal(principal);
+        if (Boolean.FALSE.equals(equalsEmailOrConfirm(request.oldEmail(), user.getEmail()))) {
+            throw new IllegalArgumentException("Invalid old email");
+        }
+        String randomNumber = authenticationService.getRandomNumber(); //вынести метод в юзерсервисе -> getRandomNumber
+        emailSender.sendSimpleMessage(request.newEmail(), "NEW CONFIRM TOKEN", randomNumber);
+        EmailTokenInfo emailTokenInfo = new EmailTokenInfo(randomNumber);
+        confirmEmail.put(request.newEmail(), emailTokenInfo);
+        return ChangeOfEmailResponse.builder()
+                .newEmail(request.newEmail())
+                .confirmToken(randomNumber)
+                .build();
+    }
+
+    private Boolean equalsEmailOrConfirm(String request, String user) {
```

Comment:
sorry, from this method's name i can't understand its purpose. 
problem is rather in its name in general. words are hard to bind with each other.

but if i debug it a little bit, then first thing that i think is making it this way:
```java
private boolean validateEmailConfirmationRequest(EmailConfirmationRequest request) {
  String confirmationToken = request.confirmationToken();
  String newEmail = request.newEmail();
  EmailTokenInfo emailTokenInfo = emailConfirmationMap.get(newEmail);
  if (emailTokenInfo == null) {
    // TODO: consider having custom exception here:
    throw new IllegalArgumentException("Invalid confirm Token"); 
  }

  return confirmationToken.equals(emailTokenInfo.getToken());
}
```